core/src/ast/attrs.rs

```diff
@@ -1,10 +1,201 @@
+            Some(Attr::DiplomatAttr(
+                a.parse_args()
+                    .expect("Failed to parse malformed diplomat::attr"),
+            ))
+        } else {
+            None
+        }
+    })
+}
+
+impl Serialize for Attrs {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: Serializer,
+    {
+        // 3 is the number of fields in the struct.
+        let mut state = serializer.serialize_struct("Attrs", 1)?;
+        let cfg: Vec<_> = self
+            .cfg
+            .iter()
+            .map(|a| quote::quote!(#a).to_string())
+            .collect();
+        state.serialize_field("cfg", &cfg)?;
+        state.end()
+    }
+}
+
+/// A `#[diplomat::attr(...)]` attribute
+///
+/// Its contents must start with single element that is a CFG-expression
+/// (so it may contain `foo = bar`, `foo = "bar"`, `ident`, `*` atoms,
+/// and `all()`, `not()`, and `any()` combiners), and then be followed by one
+/// or more backend-specific attributes, which can be any valid meta-item
+#[derive(Clone, PartialEq, Eq, Hash, Debug, Serialize)]
+pub struct DiplomatAttr {
+    pub cfg: DiplomatAttrCfg,
+    #[serde(serialize_with = "serialize_meta")]
+    pub attr: Meta,
+}
+
+fn serialize_meta<S>(m: &Meta, s: S) -> Result<S::Ok, S::Error>
+where
+    S: Serializer,
+{
+    quote::quote!(#m).to_string().serialize(s)
+}
+
+#[derive(Clone, PartialEq, Eq, Hash, Debug, Serialize)]
+pub enum DiplomatAttrCfg {
+    Not(Box<DiplomatAttrCfg>),
+    Any(Vec<DiplomatAttrCfg>),
+    All(Vec<DiplomatAttrCfg>),
+    Star,
+    BackendName(String),
+    NameValue(String, String),
+}
+
+impl Parse for DiplomatAttrCfg {
+    fn parse(input: ParseStream<'_>) -> syn::Result<Self> {
+        let lookahead = input.lookahead1();
+        if lookahead.peek(Ident) {
+            let name: Ident = input.parse()?;
+            if name == "not" {
+                let content;
+                let _paren = syn::parenthesized!(content in input);
+                Ok(DiplomatAttrCfg::Not(Box::new(content.parse()?)))
+            } else if name == "any" || name == "all" {
+                let content;
+                let _paren = syn::parenthesized!(content in input);
+                let list = content.parse_terminated(Self::parse, Token![,])?;
+                let vec = list.into_iter().collect();
+                if name == "any" {
+                    Ok(DiplomatAttrCfg::Any(vec))
+                } else {
+                    Ok(DiplomatAttrCfg::All(vec))
+                }
+            } else if input.peek(Token![=]) {
+                let _t: Token![=] = input.parse()?;
+                if input.peek(Ident) {
+                    let value: Ident = input.parse()?;
+                    Ok(DiplomatAttrCfg::NameValue(
+                        name.to_string(),
+                        value.to_string(),
+                    ))
+                } else {
+                    let value: LitStr = input.parse()?;
```

Comment:
Thought: Ident and LitStr are probably the two main cases here, but is there a reason we don't just allow a syn::Expr, like in syn::MetaNameValue?

Reply:
Don't need it yet, and want to simplify parsing later. If we do a `syn::Expr` then we need to write code to evaluate it